File: leetcode problems/Construct String from Binary Tree.cpp
class Solution
{
public:
    string solve(TreeNode *root)
    {
        if (root == NULL)
            return "";

        if (root->left == NULL && root->right == NULL)
            return to_string(root->val) + "";
        if (root->right == NULL)
            return to_string(root->val) + "(" + solve(root->left) + ")";
        return to_string(root->val) + "(" + solve(root->left) + ")" + "(" + solve(root->right) + ")";
    }

    string tree2str(TreeNode *root)
    {
        return solve(root);
    }
};

File: Construct String from Binary Tree.py
class Solution:
    def tree2str(self, root: Optional[TreeNode]) -> str:
        def preorder(node):
            if not node:
                return ""
            
            result = str(node.val)
            
            if node.left or node.right:
                result += "(" + preorder(node.left) + ")"
            
            if node.right:
                result += "(" + preorder(node.right) + ")"
            
            return result
        
        return preorder(root)
